poc_mobile_verifier/src/speedtests.rs

```diff
@@ -0,0 +1,240 @@
+            "#,
+        )
+        .bind(self.id)
+        .bind(self.speedtests)
+        .fetch_one(exec)
+        .await
+        .map(|result| result.inserted)
+        .map_err(Error::from)
+    }
+}
+
+#[derive(Clone)]
+pub struct SpeedtestAverages {
+    // I'm not sure that VecDeque is actually all that useful here, considering
+    // we have to constantly convert between the two.
+    // It does make me more confident in the implementation of validate_speedtests
+    // though.
+    pub speedtests: HashMap<PublicKey, VecDeque<Speedtest>>,
+}
+
+impl SpeedtestAverages {
+    pub fn into_iter(self) -> impl IntoIterator<Item = SpeedtestRollingAverage> {
+        self.speedtests
+            .into_iter()
+            .map(|(id, window)| SpeedtestRollingAverage {
+                id,
+                speedtests: Vec::from(window),
+            })
+    }
+
+    pub fn get_average(&self, pub_key: &PublicKey) -> Option<Average> {
+        self.speedtests.get(pub_key).map(Average::from)
+    }
+
+    pub async fn validated(
+        exec: impl sqlx::PgExecutor<'_> + Copy,
+        starting: DateTime<Utc>,
+    ) -> std::result::Result<Self, sqlx::Error> {
+        let mut speedtests = HashMap::new();
+
+        let mut rows = sqlx::query_as::<_, SpeedtestRollingAverage>(
+            "SELECT * FROM speedtests where timestamp >= $1",
+        )
+        .bind(starting.naive_utc())
+        .fetch(exec);
+
+        while let Some(SpeedtestRollingAverage {
+            id,
+            speedtests: window,
+        }) = rows.try_next().await?
+        {
+            speedtests.insert(id, VecDeque::from(window));
+        }
+
+        Ok(Self { speedtests })
+    }
+
+    pub async fn validate_speedtests(
+        exec: impl sqlx::PgExecutor<'_> + Copy,
+        file_store: &FileStore,
+        epoch: &Range<DateTime<Utc>>,
+    ) -> Result<Self> {
+        let mut speedtests = HashMap::new();
+
+        for CellSpeedtest {
+            pubkey,
+            timestamp,
+            upload_speed,
+            download_speed,
+            latency,
+            ..
+        } in crate::ingest::new_speedtest_reports(file_store, epoch)
+            .await?
+            .into_iter()
+        {
+            if !speedtests.contains_key(&pubkey) {
+                let SpeedtestRollingAverage {
+                    id,
+                    speedtests: window,
+                } = SpeedtestRollingAverage::fetch(exec, &pubkey)
+                    .await
+                    .unwrap_or_else(|_| SpeedtestRollingAverage::new(pubkey.clone(), Vec::new()));
+
+                speedtests.insert(id, VecDeque::from(window));
+            }
+
+            let timestamp = timestamp.naive_utc();
+            let upload_speed = upload_speed as i64;
+            let download_speed = download_speed as i64;
+            let latency = latency as i32;
+
+            // Unwrap here is guaranteed never to panic
+            let window = speedtests.get_mut(&pubkey).unwrap();
```

Comment:
Couldn't you use `speedtests.entry(&pubkey).or_insert_with(...)` above and then avoid this statement all together?

Reply:
Unfortunately not, as the closure would need to be async